File: questao7/python/associacao.py
# Definindo a classe Professor
class Professor:
    def __init__(self, nome):
        self.nome = nome
        self.escolas = []  # Lista de escolas onde o professor leciona

    def adicionar_escola(self, escola):
        if escola not in self.escolas:
            self.escolas.append(escola)
            escola.adicionar_professor(self)  # Associação bidirecional

    def listar_escolas(self):
        escolas_nomes = [escola.nome for escola in self.escolas]
        return f"Professor {self.nome} leciona em: {', '.join(escolas_nomes)}"


# Definindo a classe Escola
class Escola:
    def __init__(self, nome):
        self.nome = nome
        self.professores = []  # Lista de professores associados à escola

    def adicionar_professor(self, professor):
        if professor not in self.professores:
            self.professores.append(professor)

    def listar_professores(self):
        professores_nomes = [professor.nome for professor in self.professores]
        return f"Escola {self.nome} tem os seguintes professores: {', '.join(professores_nomes)}"


# Criando escolas
escola1 = Escola("Escola A")
escola2 = Escola("Escola B")

# Criando professores
professor1 = Professor("Prof. João")
professor2 = Professor("Prof. Maria")

# Associando professores às escolas
professor1.adicionar_escola(escola1)
professor1.adicionar_escola(escola2)
professor2.adicionar_escola(escola1)

# Listando professores em cada escola
print(escola1.listar_professores())
print(escola2.listar_professores())

# Listando escolas em que cada professor leciona
print(professor1.listar_escolas())
print(professor2.listar_escolas())

File: questao7/go/associacao.go
package main

import "fmt"

// Struct Professor
type Professor struct {
    nome    string
    escolas []*Escola  // Lista de escolas onde o professor leciona
}

// Método para associar uma escola ao professor
func (p *Professor) adicionarEscola(escola *Escola) {
    p.escolas = append(p.escolas, escola)
    escola.adicionarProfessor(p)  // Associação bidirecional
}

// Método para listar escolas onde o professor leciona
func (p Professor) listarEscolas() string {
    resultado := fmt.Sprintf("Professor %s leciona em: ", p.nome)
    for _, escola := range p.escolas {
        resultado += escola.nome + ", "
    }
    return resultado
}

// Struct Escola
type Escola struct {
    nome       string
    professores []*Professor  // Lista de professores na escola
}

// Método para associar um professor à escola
func (e *Escola) adicionarProfessor(professor *Professor) {
    e.professores = append(e.professores, professor)
}

// Método para listar professores da escola
func (e Escola) listarProfessores() string {
    resultado := fmt.Sprintf("Escola %s tem os seguintes professores: ", e.nome)
    for _, professor := range e.professores {
        resultado += professor.nome + ", "
    }
    return resultado
}

func main() {
    // Criando escolas
    escola1 := &Escola{nome: "Escola A"}
    escola2 := &Escola{nome: "Escola B"}

    // Criando professores
    professor1 := &Professor{nome: "Prof. João"}
    professor2 := &Professor{nome: "Prof. Maria"}

    // Associando professores às escolas
    professor1.adicionarEscola(escola1)
    professor1.adicionarEscola(escola2)
    professor2.adicionarEscola(escola1)

    // Listando professores em cada escola
    fmt.Println(escola1.listarProfessores())
    fmt.Println(escola2.listarProfessores())

    // Listando escolas onde cada professor leciona
    fmt.Println(professor1.listarEscolas())
    fmt.Println(professor2.listarEscolas())
}
